### PR TITLE
Header User Conditionals / Additionals

### PR DESCRIPTION
**Pull request by daviddyess**
- Header conditionals:
  - Hide Login/Register if user is logged in
  - Display User if user is logged in
- Added Dashboard link to User (will require a role check in the future)
- Added Logout link to User

### CODE REVIEW

File: src/components/Header/Header.js [hunk 1]
```diff
@@ -48,22 +71,41 @@ export default class Header extends Component {
...
+                        onClick={this.logoutUser}
+                      >
+                        Logout
+                        <br />
+                      </NavDropdown.Item>
+                    </NavDropdown>
+                  ) : (
+                    ''
+                  )}
+                  {!loggedIn ? this.renderNavItem('/login', 'Login') : ''}
```
Comment (ayan4m1) [inline src/components/Header/Header.js]: Might want to wrap these two in a `<Fragment>` so you only need one ternary.

File: src/sagas/application.js [hunk 2]
```diff
@@ -172,6 +172,39 @@ function* loginUserWorker({ emailAddress, password }) {
   }
 }
 
+function* logoutUserWorker() {
+  try {
+    // first, check to see if we are logged in
+    const loggedIn = yield select(isLoggedIn);
+
+    if (loggedIn) {
+      localStorage.setItem('accessToken', null);
```
Comment (ayan4m1) [inline src/sagas/application.js]: Probably want `localStorage.removeItem` here - [docs](https://developer.mozilla.org/en-US/docs/Web/API/Storage/removeItem)

File: src/sagas/application.js [hunk 3]
```diff
@@ -172,6 +172,39 @@ function* loginUserWorker({ emailAddress, password }) {
...
+      localStorage.setItem('expiration', null);
+    }
+
```
Comment (ayan4m1) [inline src/sagas/application.js]: Don't we need to clear the accessToken and expiration from redux state as well?

File: src/components/Header/Header.js (hunk 1 above)
Comment (daviddyess) [reply src/components/Header/Header.js]: This has been implemented

File: src/sagas/application.js (hunk 3 above)
Comment (daviddyess) [reply src/sagas/application.js]: logoutUserSuccess resets the 'authorization' state to initialState

File: src/sagas/application.js (hunk 2 above)
Comment (daviddyess) [reply src/sagas/application.js]: Refactored to `localStorage.removeItem()`.  I had read somewhere that `localStorage.setItem()` is better for performance than deleting and creating, but I can't find the reference; in this case I doubt there it makes a difference one way or the other.

File: src/components/Header/Header.js [hunk 4]
```diff
@@ -104,8 +104,14 @@ export class Header extends Component {
...
                   )}
-                  {!loggedIn ? this.renderNavItem('/login', 'Login') : ''}
-                  {!loggedIn ? this.renderNavItem('/register', 'Register') : ''}
+                  {!loggedIn ? (
+                    <Fragment>
+                      {this.renderNavItem('/login', 'Login')}
+                      {this.renderNavItem('/register', 'Register')}
+                    </Fragment>
+                  ) : (
+                    ''
```
Comment (ayan4m1) [inline src/components/Header/Header.js]: Can this be `null` instead of an empty string?

Comment (daviddyess) [reply src/components/Header/Header.js]: Updated

**Comment by daviddyess** [comment]
@ayan4m1 any other change requests for this PR?

File: src/components/Header/Header.test.js [hunk 5]
```diff
@@ -1,18 +1,69 @@
...
+  it('renders logged out correctly', () => {
+    const props = {
+      loggedIn: false
+    };
+
+    expect(
+      renderer
+        .create(
+          <Provider store={store}>
+            <RoutedHeader actions={actions} {...props} />
```
Comment (ayan4m1) [inline src/components/Header/Header.test.js]: If this is the inner, unconnected Header export, then there should not be a need for a `<Provider>` - if `Header` itself uses/references a connected component, that can be mocked like

```
jest.mock('components/SomeComponent/SomeComponent', () =>
  require('utils').mockComponent('SomeComponent')
);
```

File: src/components/Header/__snapshots__/Header.test.js.snap [hunk 6]
```diff
@@ -170,8 +170,127 @@ exports[`<Header /> renders correctly 1`] = `
...
+              aria-current={null}
+              className="nav--link-custom px-3 nav-link"
+              data-rb-event-key={null}
+              disabled={false}
+              href="/calculator"
+              onClick={[Function]}
+              role="tab"
+              tabIndex={-1}
+            >
+              Calculator
```
Comment (ayan4m1) [inline src/components/Header/__snapshots__/Header.test.js.snap]: Why is the logged-out snapshot showing items that shouldn't render for it?

Comment (daviddyess) [reply src/components/Header/__snapshots__/Header.test.js.snap]: I added a check for Calculator, are there any other items that don't need to be displayed?

File: src/components/Header/Header.test.js [hunk 7]
```diff
@@ -15,50 +11,33 @@ describe('<Header />', () => {
...
 
     expect(
-      renderer
-        .create(
-          <Provider store={store}>
-            <RoutedHeader actions={actions} {...props} />
-          </Provider>
-        )
-        .toJSON()
+      renderer.create(<RoutedHeader {...props} />).toJSON()
```
Comment (ayan4m1) [inline src/components/Header/Header.test.js]: 👏

**Pull request merged**